File: goldencheck/profilers/range_distribution.py
"""Range and distribution profiler — detects outliers and reports min/max for numeric columns."""
from __future__ import annotations
import polars as pl
from goldencheck.models.finding import Finding, Severity
from goldencheck.profilers.base import BaseProfiler

NUMERIC_DTYPES = (
    pl.Int8, pl.Int16, pl.Int32, pl.Int64,
    pl.UInt8, pl.UInt16, pl.UInt32, pl.UInt64,
    pl.Float32, pl.Float64,
)


class RangeDistributionProfiler(BaseProfiler):
    def profile(self, df: pl.DataFrame, column: str, *, context: dict | None = None) -> list[Finding]:
        findings: list[Finding] = []
        col = df[column]
        dtype = col.dtype
        is_numeric = dtype in NUMERIC_DTYPES

        # Chain: if type inference flagged as mostly numeric, cast and run
        if not is_numeric and context and context.get(column, {}).get("mostly_numeric"):
            col = col.cast(pl.Float64, strict=False).drop_nulls()
            is_numeric = True
        elif not is_numeric:
            return findings

        non_null = col.drop_nulls() if is_numeric and dtype in NUMERIC_DTYPES else col
        total = len(non_null)
        if total < 2:
            return findings

        mean = non_null.mean()
        std = non_null.std()
        col_min = non_null.min()
        col_max = non_null.max()

        findings.append(Finding(
            severity=Severity.INFO,
            column=column,
            check="range_distribution",
            message=f"Range: min={col_min}, max={col_max}, mean={mean:.2f}",
        ))

        if std is not None and std > 0:
            lower = mean - 3 * std
            upper = mean + 3 * std
            outliers = non_null.filter((non_null < lower) | (non_null > upper))
            outlier_count = len(outliers)
            if outlier_count > 0:
                sample = outliers.head(5).to_list()
                # Determine how many stddevs outliers are
                # Use max deviation to determine confidence
                max_dev = max(
                    abs(float(non_null.max()) - mean) / std,
                    abs(float(non_null.min()) - mean) / std,
                ) if std > 0 else 0
                confidence = 0.9 if max_dev > 5 else 0.7
                findings.append(Finding(
                    severity=Severity.WARNING,
                    column=column,
                    check="range_distribution",
                    message=f"{outlier_count} outlier(s) detected beyond 3 standard deviations",
                    affected_rows=outlier_count,
                    sample_values=[str(v) for v in sample],
                    suggestion="Investigate outlier values for data entry errors or anomalies",
                    confidence=confidence,
                ))

        return findings

File: packages/goldencheck-js/src/core/profilers/range-distribution.ts
/**
 * Range and distribution profiler — detects outliers and reports min/max.
 * Port of goldencheck/profilers/range_distribution.py.
 */

import type { TabularData } from "../data.js";
import { mean as statMean, std as statStd } from "../stats.js";
import { type Finding, Severity, makeFinding } from "../types.js";
import type { Profiler } from "./base.js";

export class RangeDistributionProfiler implements Profiler {
  profile(
    data: TabularData,
    column: string,
    context?: Record<string, unknown>,
  ): Finding[] {
    const findings: Finding[] = [];

    let isNumeric = data.isNumeric(column);

    // Chain: if type inference flagged as mostly numeric, treat as numeric
    if (
      !isNumeric &&
      context &&
      (context[column] as Record<string, unknown> | undefined)?.["mostly_numeric"]
    ) {
      isNumeric = true;
    }

    if (!isNumeric) return findings;

    const nums = data.numericValues(column);
    if (nums.length < 2) return findings;

    const m = statMean(nums)!;
    const s = statStd(nums);
    const colMin = Math.min(...nums);
    const colMax = Math.max(...nums);

    findings.push(
      makeFinding({
        severity: Severity.INFO,
        column,
        check: "range_distribution",
        message: `Range: min=${colMin}, max=${colMax}, mean=${m.toFixed(2)}`,
      }),
    );

    if (s !== null && s > 0) {
      const lower = m - 3 * s;
      const upper = m + 3 * s;
      const outliers = nums.filter((v) => v < lower || v > upper);
      const outlierCount = outliers.length;

      if (outlierCount > 0) {
        const sample = outliers.slice(0, 5).map(String);
        const maxDev = Math.max(
          Math.abs(colMax - m) / s,
          Math.abs(colMin - m) / s,
        );
        const confidence = maxDev > 5 ? 0.9 : 0.7;

        findings.push(
          makeFinding({
            severity: Severity.WARNING,
            column,
            check: "range_distribution",
            message: `${outlierCount} outlier(s) detected beyond 3 standard deviations`,
            affectedRows: outlierCount,
            sampleValues: sample,
            suggestion: "Investigate outlier values for data entry errors or anomalies",
            confidence,
          }),
        );
      }
    }

    return findings;
  }
}
